README.md
shorty
======

This repository contains the main 'Shorty' owncloud app and one or more plugins to that app. 

Shorty itself is a comfortable url shortening suite with a twist. 

The plugin(s) provide additional features but are standalone apps from a technical point of view: 

shorty_tracking: tracking of requests and details to existing entries in the main app


Folder layout: 
- shorty:          the main "Shorty" app
- shorty_tracking: the "Shorty Tracking" plugin to the main app



doc/CHANGELOG
*****
* @package shorty an ownCloud url shortener plugin
* @category internet
* @author Christian Reiner
* @copyright 2011-2013 Christian Reiner <foss@christian-reiner.info>
* @license GNU Affero General Public license (AGPL)
* @link information http://apps.owncloud.com/content/show.php/Shorty?content=150401
* @link repository https://svn.christian-reiner.info/svn/app/oc/shorty
*
* This library is free software; you can redistribute it and/or
* modify it under the terms of the GNU AFFERO GENERAL PUBLIC LICENSE
* License as published by the Free Software Foundation; either
* version 3 of the license, or any later version.
*
* This library is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU AFFERO GENERAL PUBLIC LICENSE for more details.
*
* You should have received a copy of the GNU Affero General Public
* License along with this library.
* If not, see <http://www.gnu.org/licenses/>.
*****

Changelog: 

* Sun Mar 24 2013 Christian Reiner: version 0.3.24
Fixed shortlet not to result in an error when being used.
* Wed Mar 20 2013 Christian Reiner: version 0.3.23
Allow direct activation in ownCloud5 based installations. 
* Fri Mar 01 2013 Christian Reiner: version 0.3.22
Some general security fix to the apps framework.
* Wed Feb 20 2013 Christian Reiner: version 0.3.21
Fixed visibility of list header which was accidentially hidden below the control panel.
Fixed display of feedback on actions by reimplementing the messenger area. 
Added a new preference to control the level of verbosity of feedback on actions.
Optimized preference retrieval (caching strategy).
* Tue Feb 15 2013 Christian Reiner: version 0.3.20
Fixed decoding of titles when scraping meta data from target pages. 
* Tue Feb 12 2013 Christian Reiner: version 0.3.19
Support for ownCloud version 5. 
Fixed handling of favicons when adding multiple Shortys. 
Update the favicon when a Shortys target url is changed. 
Enhancement to the list layout and apeparance to improve readability. 
* Mon Jan 14 2013 Christian Reiner: version 0.3.18
Fixed the 'recycle' icon for rows marked as deleted, to not scrumble row layout.
Switched more (dynamically controlled) icons to svg format.
Fixed some problems in compatibility to ownCloud version 4.0 based installations.
Action icons now stay visible inside the active row when not hovering.
Size optimization for the png versions of icons.
* Mon Jan 07 2013 Christian Reiner: version 0.3.17
Use svg version of icons where applicable.
Recreated most icons for a more unified look and feel.
Fixed qrcode retrieval uder PostgreSQL database engines.
Generally prevent text and content selections except for input elements.
Numerous fixes to get Shorty usable under MS Internet Explorer.
* Tue Jan 01 2013 Christian Reiner: version 0.3.16
Fixed automatic text selection inside the usage popups.
Force downloading of a qrcode image when self referencing it.
* Mon Dec 31 2012 Christian Reiner: version 0.3.15
Changed supported OC version to 4.9 (OC-4.5).
Support for the postgresql database engine.
The apps top controls panel can be hidden now and got an enhanced layout.
Some style and layout corrections especially for webkit based browsers.
Fixed validation of the static backend setup for cross domain situations.
Better support for cross domain verification for the static backend configuration.
* Sat Dec 15 2012 Christian Reiner: version 0.3.14
Fixed big oops display of access times that gave the wrong month and day!
* Wed Dec 11 2012 Christian Reiner: version 0.3.13
Fixed layout of share dialog getting distorted under certain circumstances. 
* Thu Nov 29 2012 Christian Reiner: version 0.3.12
Fixed a small glitch in settings handling. 
Some minor performance and logging enhancements.
* Fri Nov 16 2012 Christian Reiner: version 0.3.11
Fixed visualization of last access time inside Shorty dialogs.
Enabled forwarding to more http based protocol schemes: shttp, ipp, webdav, webdavs
* Tue Nov 01 2012 Christian Reiner: version 0.3.10
Some fixed to the actions inside dialogs when used inside OC-4.0.x installations. 
An additional service 'shorty_stats' allows to remotely retrieve data.
* Fri Oct 26 2012 Christian Reiner: version 0.3.9
Targets of existing Shortys can be modified as an explicit action now.
Enhanced notification and exception visualization and handling.
A few bugfixes and enhancements for title handling inside Shortys.
Some further layout and style fixes, cleanups. 
* Tue Oct 09 2012 Christian Reiner: version 0.3.8
Switched layout of list of shortys to a liquid style (using the whole screens width).
Fixed display size for some exotic favicons inside dialogs.
Internal style reorganisation to isolate OC4 backwards compatibility exceptions.
Fixed visualization of explanations for goo.gl backend in preferences dialog.
Fixed sharpness control of dialogs.
Fixed datepicker dialog sometimes being hidden by other widgets.
Some fixes for usage under MS Internet Explorer though still a lot of issues remain.
A few layout enhancements and german translation completion.
* Fri Sep 28 2012 Christian Reiner: version 0.3.7
Preparations to rely on the OC core to refresh the request token if implemented.
- the Shorty strategy and code is to be accepted into the OC core :-)
- for version prior to that Shorty still handles refreshing itself
Fixed a crash that was introduced by changes in the OC core in recent beta versions.
Some preparations for future referencing of files stored inside OC as opposed to URLs.
Some internal code changes to reflect advances in the underlying jquery library.
A few logical fixes to detection of special runtime exceptions. 
Some more small optical layout fixes. 
Fixed missleading debug statement in ajax hooks for settings and preferences control
Fixed settings and preferences control not to ignore empty values
* Sat Sep 09 2012 Christian Reiner: version 0.3.6

index.php
<?php
/**
* @package shorty an ownCloud url shortener plugin
* @category internet
* @author Christian Reiner
* @copyright 2011-2013 Christian Reiner <foss@christian-reiner.info>
* @license GNU Affero General Public license (AGPL)
* @link information http://apps.owncloud.com/content/show.php/Shorty?content=150401
* @link repository https://svn.christian-reiner.info/svn/app/oc/shorty
*
* This library is free software; you can redistribute it and/or
* modify it under the terms of the GNU AFFERO GENERAL PUBLIC LICENSE
* License as published by the Free Software Foundation; either
* version 3 of the license, or any later version.
*
* This library is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU AFFERO GENERAL PUBLIC LICENSE for more details.
*
* You should have received a copy of the GNU Affero General Public
* License along with this library.
* If not, see <http://www.gnu.org/licenses/>.
*
*/

/**
 * @file index.php
 * This is the plugins central position
 * All requests to the plugin are handled by this file.
 * Exceptions: system settings, user preferences and relaying
 * @access public
 * @author Christian Reiner
 */

// Session checks
OCP\User::checkLoggedIn  ( );
OCP\App::checkAppEnabled ( 'shorty' );

OCP\App::setActiveNavigationEntry ( 'shorty_index' );

OCP\Util::addStyle  ( 'shorty/3rdparty', 'jquery-impromptu' );
OCP\Util::addStyle  ( 'shorty',          'shorty' );
// TODO: remove OC-4.0-compatibility:
if (OC_Shorty_Tools::versionCompare('<','4.80')) // OC-4.0
	OCP\Util::addStyle ( 'shorty', 'shorty-oc40' );
// TODO: remove OC-4.5-compatibility:
if (OC_Shorty_Tools::versionCompare('<','4.91')) // OC-4.5
	OCP\Util::addStyle ( 'shorty', 'shorty-oc45' );

OCP\Util::addscript ( '3rdparty',        'chosen/chosen.jquery.min');
OCP\Util::addScript ( 'shorty/3rdparty', 'jquery-tinysort.min' );
OCP\Util::addScript ( 'shorty/3rdparty', 'jquery-impromptu' );
OCP\Util::addScript ( 'shorty',          'shorty' );
OCP\Util::addScript ( 'shorty',          'util' );
OCP\Util::addScript ( 'shorty',          'init' );
if ( OC_Log::DEBUG==OC_Config::getValue( "loglevel", OC_Log::WARN ) )
	OCP\Util::addScript ( 'shorty',  'debug' );
// any additional stuff to incude as registered into the hook ?
OC_Shorty_Hooks::requestIncludes();

// strategy:
// - first: decide which action is requested
// - second: execute that action with an optional argument provided

// defaults:
$act = 'index';
$arg = NULL;
// we try to guess what the request indicates:
// - a (shorty) id to be looked up in the database resulting in a forwarding to the stored target
// - a (target) url to be added as a new shorty
// - none of the two, so just a plain list of existing shortys
foreach ($_GET as $key=>$val) // in case there are unexpected, additional arguments like a timestamp added by some stupid proxy
{
	switch ($key)
	{
		// this is the OC4 argument used to identify the app called, we ignore it:
		case 'app':
			break;

		// any recognizable argument key indicating a file to be added as new shorty ?
		case 'file':
		case 'dir':
		case 'folder':
			// example: http://.../shorty/index.php?file=%3cfolder%3cfile
			$act = 'acquire';
			$arg = OC_Shorty_Type::req_argument($key,OC_Shorty_Type::PATH,FALSE);
			break 2; // skip switch AND foreach

		// any recognizable argument key indicating a url to be added as new shorty ?
		case 'url':
		case 'uri':
		case 'link':
			// example: http://.../shorty/index.php?url=http%...
			$act = 'acquire';
			$arg = OC_Shorty_Type::req_argument($key,OC_Shorty_Type::URL,FALSE);
			break 2; // skip switch AND foreach

		// no recognizable key but something else, hm...
		// this _might_ be some unexcepted argument, or:
		// it is an expected argument, but without recognizable key, so we try to guess by examining the content
		// we restrict this 'guessing' to cases where only a single argument is specified
		default:
			if (  (1==sizeof($_GET))  // only one single request argument
				&&( ! reset($_GET)) ) // no value, so maybe just an id
			{
				// use that source instead of $key, since $key contains replaced chars (php specific exceptions due to var name problems)
				$raw = urldecode($_SERVER['QUERY_STRING']);
				// now try to interpret its content
				if (NULL!==($value=OC_Shorty_Type::normalize($raw,OC_Shorty_Type::URL,FALSE)))
				{
					// the query string is a url, acquire it as a new shorty
					$act = 'acquire';
					$arg = $raw;
					break 2;
				}
				else
				{
					// no pattern recognized, so we assume an ordinary index action
					$act = 'index';
					break 2;
				}
			} // if
			$act='index';
			break 2;
	} // switch key
} // foreach key

// next, execute the "act" whilst considering the 'arg'
switch ($act)
{
	case 'acquire': // add url as new shorty
		// keep the url specified as referer, that is the one we want to store
		$_SESSION['shorty-referrer'] = $arg;
		OCP\Util::writeLog( 'shorty', sprintf("Detected an incoming Shortlet request for url '%s...'",substr($arg,0,80)), OC_Log::DEBUG );
		header ( sprintf('Location: %s', OCP\Util::linkTo('shorty','index.php')) );
		exit();

	// =====
	case 'index': // action 'index': list of shortys
	default:
		try
		{
			// is this a redirect from a call with a target url to be added ?
			if ( isset($_SESSION['shorty-referrer']) )
			{
				// this takes care of handling the url on the client side
				OCP\Util::addScript ( 'shorty', 'add' );
				// add url taked from the session vars to anything contained in the query string
				$_SERVER['QUERY_STRING'] = implode('&',array_merge(array('url'=>$_SESSION['shorty-referrer']),explode('&',$_SERVER['QUERY_STRING'])));
			}
			else
			{
				// simple desktop initialization, no special actions contained
				OCP\Util::addScript ( 'shorty', 'list' );
			}
			$tmpl = new OCP\Template( 'shorty', 'tmpl_index', 'user' );
			// any additional actions registered via hooks that should be offered ?
			$tmpl->assign ( 'shorty-actions', OC_Shorty_Hooks::requestActions() );
			// the (remote) base url of the qrcode generator
			$tmpl->assign ( 'qrcode-ref', sprintf('%s?service=%s&id=',  OCP\Util::linkToAbsolute('', 'public.php'), 'shorty_qrcode') );
			// available status options (required for select filter in toolbox)
			$shorty_status['']=sprintf('- %s -',OC_Shorty_L10n::t('all'));
			foreach ( OC_Shorty_Type::$STATUS as $status )
				$shorty_status[$status] = OC_Shorty_L10n::t($status);
			$tmpl->assign ( 'shorty-status', $shorty_status );
			// any referrer we want to hand over to the browser ?
			if ( array_key_exists('shorty-referrer',$_SESSION) )
				$tmpl->assign ( 'shorty-referrer', $_SESSION['shorty-referrer'] );
			// is sending sms enabled in the personal preferences ?
			$tmpl->assign ( 'sms-control', OCP\Config::getUserValue(OCP\User::getUser(),'shorty','sms-control','disabled') );
			// clean up session var so that a browser reload does not trigger the same action again
			unset ( $_SESSION['shorty-referrer'] );
			$tmpl->printPage();
		} catch ( OC_Shorty_Exception $e ) { OCP\JSON::error ( array ( 'message'=>$e->getTranslation(), 'level'=>'error', 'data'=>$result ) ); }
} // switch
?>
